With project references, some cross-package imports of inferred types are emitted with relative paths

<!-- 🚨 STOP 🚨 STOP 🚨 STOP 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!--
Please try to reproduce the issue with the latest published version. It may have already been fixed.
For npm: `typescript@next`
This is also the 'Nightly' version in the playground: http://www.typescriptlang.org/play/?ts=Nightly
-->
**TypeScript Version:**  3.9.5 (also repros with nightly)

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** project references, declaration files, tsconfig paths, inferred types

**Code**

The repro requires a monorepo with project references plus `paths` in the tsconfig. Full repro with the problematic compiled output checked in here: https://github.com/ecraig12345/learn-a/tree/inferred-relative-imports (it's fairly minimal but may still include a few settings which turn out to be tangential to the issue, sorry)

Here's a summary...

`@fluentui/pkg2` depends on `@fluentui/pkg1`, and both live in the same monorepo under a `packages` folder.

[`pkg2`'s tsconfig](https://github.com/ecraig12345/learn-a/blob/inferred-relative-imports/packages/pkg2/tsconfig.json) has a project reference pointing to `pkg1`, and both build with `composite: true`.

They inherit from a [shared tsconfig](https://github.com/ecraig12345/learn-a/blob/inferred-relative-imports/tsconfig.json) which has a path config like this (to allow imports directly from the root of each package):
```
"baseUrl": ".",
"paths": {
  "@fluentui/*": ["packages/*/src"]
}
```

The problem comes when you have a file in one package ([`pkg2/src/index.ts`](https://github.com/ecraig12345/learn-a/blob/inferred-relative-imports/packages/pkg2/src/index.ts) in the repro) which has an implicit/inferred reference to a type from another local project-referenced package. 
```ts
import { IThings } from '@fluentui/pkg1';

export function fn4() { // inferred return type: IThings from pkg1
  const a: IThings = { thing1: { a: 'b' } };
  return a.thing1;
}
```
This inferred type need not be stated locally, but it will show up in the declaration file as an inline `import()`.

**Expected behavior:**

Compile each package with `tsc` (or in the repro, run `yarn build` from the root).

`pkg2`'s `index.d.ts` looks like this:
```ts
export declare function fn4(): import("@fluentui/pkg1").IThing;
```

**Actual behavior:**

[`pkg2`'s `index.d.ts`](https://github.com/ecraig12345/learn-a/blob/inferred-relative-imports/packages/pkg2/lib/src/index.d.ts) looks like this, which will break consumers:
```ts
export declare function fn4(): import("../../pkg1/src").IThing;
```

`pkg2/src/index.ts` output also adds an `src` folder under `lib`, which I wouldn't expect (`pkg2/lib/src/index.{js,d.ts}`), but this is probably unrelated.

Variants:
- Same behavior if `-b` is added to the build command.
- Same behavior if using `rootDir` instead of `include`, just different output directory structure.
- Same behavior if the config inheritance is removed and `paths` config is moved to each project.
- Without project references, you end up with a directory structure like this, which I also don't understand, but is not the same issue:
```
pkg2/
  lib/
    pkg1/src/<files>
    pkg2/src/<files>
```

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> can't be shown on playground

**Related Issues:** <!-- Did you find other bugs that looked similar? --> didn't see anything else quite like this

I can reproduce this bug after upgrading TypeScript from version `3.7` to `3.9.7`.

I have monorepo with:
`packages/std-react` using `packages/std-web', which is symlinked inside `packages/std-react/node_modules`

Building `packages/std-react` throws this error
```
../std-web/router.ts:77:17 - error TS2742: The inferred type of 'createHistoryRouter' cannot be named without a reference to '../std-react/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
```

but if I copy and paste `packages/std-web' inside `packages/std-react/node_modules`, the errors goes away.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -5679,7 +5679,7 @@ namespace ts {
                         specifierCompilerOptions,
                         contextFile,
                         moduleResolverHost,
-                        { importModuleSpecifierPreference: isBundle ? "non-relative" : "relative", importModuleSpecifierEnding: isBundle ? "minimal" : undefined },
+                        { importModuleSpecifierPreference: isBundle ? "non-relative" : "project-relative", importModuleSpecifierEnding: isBundle ? "minimal" : undefined },
                     ));
                     links.specifierCache ??= new Map();
                     links.specifierCache.set(contextFile.path, specifier);

--- a/src/compiler/moduleSpecifiers.ts
+++ b/src/compiler/moduleSpecifiers.ts
@@ -206,7 +206,9 @@ namespace ts.moduleSpecifiers {
         }
 
         if (relativePreference === RelativePreference.ExternalNonRelative) {
-            const projectDirectory = host.getCurrentDirectory();
+            const projectDirectory = compilerOptions.configFilePath ?
+                toPath(getDirectoryPath(compilerOptions.configFilePath), host.getCurrentDirectory(), info.getCanonicalFileName) :
+                info.getCanonicalFileName(host.getCurrentDirectory());
             const modulePath = toPath(moduleFileName, projectDirectory, getCanonicalFileName);
             const sourceIsInternal = startsWith(sourceDirectory, projectDirectory);
             const targetIsInternal = startsWith(modulePath, projectDirectory);

--- a/src/compiler/moduleSpecifiers.ts
+++ b/src/compiler/moduleSpecifiers.ts
@@ -448,7 +448,7 @@ namespace ts.moduleSpecifiers {
                         startsWith(relativeToBaseUrl, prefix) &&
                         endsWith(relativeToBaseUrl, suffix) ||
                         !suffix && relativeToBaseUrl === removeTrailingDirectorySeparator(prefix)) {
-                        const matchedStar = relativeToBaseUrl.substr(prefix.length, relativeToBaseUrl.length - suffix.length);
+                        const matchedStar = relativeToBaseUrl.substr(prefix.length, relativeToBaseUrl.length - suffix.length - prefix.length);
                         return key.replace("*", matchedStar);
                     }
                 }

--- a/src/compiler/moduleSpecifiers.ts
+++ b/src/compiler/moduleSpecifiers.ts
@@ -181,14 +181,13 @@ namespace ts.moduleSpecifiers {
     function getLocalModuleSpecifier(moduleFileName: string, info: Info, compilerOptions: CompilerOptions, host: ModuleSpecifierResolutionHost, { ending, relativePreference }: Preferences): string {
         const { baseUrl, paths, rootDirs } = compilerOptions;
         const { sourceDirectory, getCanonicalFileName } = info;
-
         const relativePath = rootDirs && tryGetModuleNameFromRootDirs(rootDirs, moduleFileName, sourceDirectory, getCanonicalFileName, ending, compilerOptions) ||
             removeExtensionAndIndexPostFix(ensurePathIsNonModuleName(getRelativePathFromDirectory(sourceDirectory, moduleFileName, getCanonicalFileName)), ending, compilerOptions);
         if (!baseUrl && !paths || relativePreference === RelativePreference.Relative) {
             return relativePath;
         }
 
-        const baseDirectory = getPathsBasePath(compilerOptions, host) || baseUrl!;
+        const baseDirectory = getNormalizedAbsolutePath(getPathsBasePath(compilerOptions, host) || baseUrl!, host.getCurrentDirectory());
         const relativeToBaseUrl = getRelativePathIfInDirectory(moduleFileName, baseDirectory, getCanonicalFileName);
         if (!relativeToBaseUrl) {
             return relativePath;

--- a/src/compiler/transformers/declarations.ts
+++ b/src/compiler/transformers/declarations.ts
@@ -372,9 +372,7 @@ namespace ts {
 
                     if (declFileName) {
                         const specifier = moduleSpecifiers.getModuleSpecifier(
-                            // We pathify the baseUrl since we pathify the other paths here, so we can still easily check if the other paths are within the baseUrl
-                            // TODO: Should we _always_ be pathifying the baseUrl as we read it in?
-                            { ...options, baseUrl: options.baseUrl && toPath(options.baseUrl, host.getCurrentDirectory(), host.getCanonicalFileName) },
+                            options,
                             currentSourceFile,
                             toPath(outputFilePath, host.getCurrentDirectory(), host.getCanonicalFileName),
                             toPath(declFileName, host.getCurrentDirectory(), host.getCanonicalFileName),

--- a/tests/baselines/reference/declarationEmitPathMappingMonorepo2.js
+++ b/tests/baselines/reference/declarationEmitPathMappingMonorepo2.js
@@ -37,5 +37,5 @@ exports["default"] = (0, utils_1.createSvgIcon)("Hello", "ArrowLeft");
 
 
 //// [index.d.ts]
-declare const _default: import("../../core/src/SvgIcon").SomeInterface;
+declare const _default: import("@ts-bug/core/SvgIcon").SomeInterface;
 export default _default;

--- a/tests/baselines/reference/declarationEmitPathMappingMonorepo2.js
+++ b/tests/baselines/reference/declarationEmitPathMappingMonorepo2.js
@@ -1,0 +1,41 @@
+//// [tests/cases/compiler/declarationEmitPathMappingMonorepo2.ts] ////
+
+//// [index.d.ts]
+export * from "./utils";
+export { default as SvgIcon } from "./SvgIcon";
+
+//// [SvgIcon.d.ts]
+import { StyledComponentProps } from "@ts-bug/styles";
+export interface SvgIconProps extends StyledComponentProps<"root"> {
+    children?: string[];
+}
+export interface SomeInterface {
+    myProp: string;
+}
+declare const SvgIcon: SomeInterface;
+export default SvgIcon;
+
+//// [utils.d.ts]
+import SvgIcon from "./SvgIcon";
+export function createSvgIcon(path: string, displayName: string): typeof SvgIcon;
+
+//// [index.d.ts]
+export interface StyledComponentProps<ClassKey extends string> {
+    classes?: Record<ClassKey, string>;
+}
+
+//// [index.ts]
+import { createSvgIcon } from "@ts-bug/core/utils";
+export default createSvgIcon("Hello", "ArrowLeft");
+
+
+//// [index.js]
+"use strict";
+exports.__esModule = true;
+var utils_1 = require("@ts-bug/core/utils");
+exports["default"] = (0, utils_1.createSvgIcon)("Hello", "ArrowLeft");
+
+
+//// [index.d.ts]
+declare const _default: import("../../core/src/SvgIcon").SomeInterface;
+export default _default;

--- a/tests/baselines/reference/declarationEmitPathMappingMonorepo2.symbols
+++ b/tests/baselines/reference/declarationEmitPathMappingMonorepo2.symbols
@@ -1,0 +1,58 @@
+=== tests/cases/compiler/packages/core/src/index.d.ts ===
+export * from "./utils";
+export { default as SvgIcon } from "./SvgIcon";
+>default : Symbol(default, Decl(SvgIcon.d.ts, 7, 37))
+>SvgIcon : Symbol(SvgIcon, Decl(index.d.ts, 1, 8))
+
+=== tests/cases/compiler/packages/core/src/SvgIcon.d.ts ===
+import { StyledComponentProps } from "@ts-bug/styles";
+>StyledComponentProps : Symbol(StyledComponentProps, Decl(SvgIcon.d.ts, 0, 8))
+
+export interface SvgIconProps extends StyledComponentProps<"root"> {
+>SvgIconProps : Symbol(SvgIconProps, Decl(SvgIcon.d.ts, 0, 54))
+>StyledComponentProps : Symbol(StyledComponentProps, Decl(SvgIcon.d.ts, 0, 8))
+
+    children?: string[];
+>children : Symbol(SvgIconProps.children, Decl(SvgIcon.d.ts, 1, 68))
+}
+export interface SomeInterface {
+>SomeInterface : Symbol(SomeInterface, Decl(SvgIcon.d.ts, 3, 1))
+
+    myProp: string;
+>myProp : Symbol(SomeInterface.myProp, Decl(SvgIcon.d.ts, 4, 32))
+}
+declare const SvgIcon: SomeInterface;
+>SvgIcon : Symbol(SvgIcon, Decl(SvgIcon.d.ts, 7, 13))
+>SomeInterface : Symbol(SomeInterface, Decl(SvgIcon.d.ts, 3, 1))
+
+export default SvgIcon;
+>SvgIcon : Symbol(SvgIcon, Decl(SvgIcon.d.ts, 7, 13))
+
+=== tests/cases/compiler/packages/core/src/utils.d.ts ===
+import SvgIcon from "./SvgIcon";
+>SvgIcon : Symbol(SvgIcon, Decl(utils.d.ts, 0, 6))
+
+export function createSvgIcon(path: string, displayName: string): typeof SvgIcon;
+>createSvgIcon : Symbol(createSvgIcon, Decl(utils.d.ts, 0, 32))
+>path : Symbol(path, Decl(utils.d.ts, 1, 30))
+>displayName : Symbol(displayName, Decl(utils.d.ts, 1, 43))
+>SvgIcon : Symbol(SvgIcon, Decl(utils.d.ts, 0, 6))
+
+=== tests/cases/compiler/packages/styles/src/index.d.ts ===
+export interface StyledComponentProps<ClassKey extends string> {
+>StyledComponentProps : Symbol(StyledComponentProps, Decl(index.d.ts, 0, 0))
+>ClassKey : Symbol(ClassKey, Decl(index.d.ts, 0, 38))
+
+    classes?: Record<ClassKey, string>;
+>classes : Symbol(StyledComponentProps.classes, Decl(index.d.ts, 0, 64))
+>Record : Symbol(Record, Decl(lib.es5.d.ts, --, --))
+>ClassKey : Symbol(ClassKey, Decl(index.d.ts, 0, 38))
+}
+
+=== tests/cases/compiler/packages/lab/src/index.ts ===
+import { createSvgIcon } from "@ts-bug/core/utils";
+>createSvgIcon : Symbol(createSvgIcon, Decl(index.ts, 0, 8))
+
+export default createSvgIcon("Hello", "ArrowLeft");
+>createSvgIcon : Symbol(createSvgIcon, Decl(index.ts, 0, 8))
+

--- a/tests/baselines/reference/declarationEmitPathMappingMonorepo2.types
+++ b/tests/baselines/reference/declarationEmitPathMappingMonorepo2.types
@@ -1,0 +1,50 @@
+=== tests/cases/compiler/packages/core/src/index.d.ts ===
+export * from "./utils";
+export { default as SvgIcon } from "./SvgIcon";
+>default : import("tests/cases/compiler/packages/core/src/SvgIcon").SomeInterface
+>SvgIcon : import("tests/cases/compiler/packages/core/src/SvgIcon").SomeInterface
+
+=== tests/cases/compiler/packages/core/src/SvgIcon.d.ts ===
+import { StyledComponentProps } from "@ts-bug/styles";
+>StyledComponentProps : any
+
+export interface SvgIconProps extends StyledComponentProps<"root"> {
+    children?: string[];
+>children : string[]
+}
+export interface SomeInterface {
+    myProp: string;
+>myProp : string
+}
+declare const SvgIcon: SomeInterface;
+>SvgIcon : SomeInterface
+
+export default SvgIcon;
+>SvgIcon : SomeInterface
+
+=== tests/cases/compiler/packages/core/src/utils.d.ts ===
+import SvgIcon from "./SvgIcon";
+>SvgIcon : import("tests/cases/compiler/packages/core/src/SvgIcon").SomeInterface
+
+export function createSvgIcon(path: string, displayName: string): typeof SvgIcon;
+>createSvgIcon : (path: string, displayName: string) => typeof SvgIcon
+>path : string
+>displayName : string
+>SvgIcon : import("tests/cases/compiler/packages/core/src/SvgIcon").SomeInterface
+
+=== tests/cases/compiler/packages/styles/src/index.d.ts ===
+export interface StyledComponentProps<ClassKey extends string> {
+    classes?: Record<ClassKey, string>;
+>classes : Record<ClassKey, string>
+}
+
+=== tests/cases/compiler/packages/lab/src/index.ts ===
+import { createSvgIcon } from "@ts-bug/core/utils";
+>createSvgIcon : (path: string, displayName: string) => import("tests/cases/compiler/packages/core/src/SvgIcon").SomeInterface
+
+export default createSvgIcon("Hello", "ArrowLeft");
+>createSvgIcon("Hello", "ArrowLeft") : import("tests/cases/compiler/packages/core/src/SvgIcon").SomeInterface
+>createSvgIcon : (path: string, displayName: string) => import("tests/cases/compiler/packages/core/src/SvgIcon").SomeInterface
+>"Hello" : "Hello"
+>"ArrowLeft" : "ArrowLeft"
+

--- a/tests/cases/compiler/declarationEmitPathMappingMonorepo2.ts
+++ b/tests/cases/compiler/declarationEmitPathMappingMonorepo2.ts
@@ -1,0 +1,44 @@
+// @filename: packages/core/src/index.d.ts
+export * from "./utils";
+export { default as SvgIcon } from "./SvgIcon";
+
+// @filename: packages/core/src/SvgIcon.d.ts
+import { StyledComponentProps } from "@ts-bug/styles";
+export interface SvgIconProps extends StyledComponentProps<"root"> {
+    children?: string[];
+}
+export interface SomeInterface {
+    myProp: string;
+}
+declare const SvgIcon: SomeInterface;
+export default SvgIcon;
+
+// @filename: packages/core/src/utils.d.ts
+import SvgIcon from "./SvgIcon";
+export function createSvgIcon(path: string, displayName: string): typeof SvgIcon;
+
+// @filename: packages/styles/src/index.d.ts
+export interface StyledComponentProps<ClassKey extends string> {
+    classes?: Record<ClassKey, string>;
+}
+
+// @filename: packages/lab/src/index.ts
+import { createSvgIcon } from "@ts-bug/core/utils";
+export default createSvgIcon("Hello", "ArrowLeft");
+
+// @filename: packages/lab/tsconfig.json
+{
+    "compilerOptions": {
+        "outDir": "dist",
+        "declaration": true,
+        "baseUrl": "../",
+        "paths": {
+            "@ts-bug/core": ["./core/src"],
+            "@ts-bug/core/*": ["./core/src/*"],
+            "@ts-bug/lab": ["./lab/src"],
+            "@ts-bug/lab/*": ["./lab/src/*"],
+            "@ts-bug/styles": ["./styles/src"],
+            "@ts-bug/styles/*": ["./styles/src/*"]
+        }
+    }
+}

